src/main/java/com/google/sps/DataGraph.java

```diff
@@ -295,15 +317,44 @@ private GraphNode changeNodeToken(GraphNode node, TokenMutation tokenMut) {
     TokenMutation.Type tokenMutType = tokenMut.getType();
     if (tokenMutType == TokenMutation.Type.ADD_TOKEN) {
       tokenList.addAll(tokenNames);
+
+      // Update the map
+      for (String tokenName : tokenNames) {
+        addToken(tokenName, node.name());
+      }
     } else if (tokenMutType == TokenMutation.Type.DELETE_TOKEN) {
       tokenList.removeAll(tokenNames);
+      // Update the map
+      for (String tokenName : tokenNames) {
+        if (this.tokenMap().containsKey(tokenName)) {
```

Comment:
I like how `addToken` is its own separate function. Maybe `removeToken` could also be a separate function?